### DISCUSSION TITLE
Using async web server with esp32 and autoconnect

### DISCUSSION THREAD

**Discussion by Tony-The-Developer** [Hieromon/AutoConnect#351] [Q&A]
I am trying to have the temperature auto-refresh. How can I use an async webserver?

**Comment by Hieromon** [comment]
This topic continues in the Discussions.

**Reply by Tony-The-Developer** [reply]
👍

**Comment by Hieromon** [comment]
ESP8266WebServer(also WebServer for ESP32) and ESPAsyncWebServer cannot coexist at the same time. If you use ESPAsyncWebServer, AutoConnect must support it too.
As a personal experiment, I implemented AsyncAutoConnect with PageBuilder corresponding to [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer). This implementation has strong restrictions and is not yet open to the public repository.

Speaking from the implementation experience, I think that you will be forced to at least the following problem to be solved.

1. We can not use delay or while loop<br>In the first place, AutoConnect does not have an asynchronous structure. It uses a delay and a While Loop to wait for a WiFi connection to be established. You need to change this structure to be event-driven. This requires changing the simple HTTP request/response model used by various AutoConnect web pages to a model that can notify events such as WebSocket.<br>In addition, it is necessary to change the wait processing by the delay loop to the thread processing that can notify interrupts such as Ticker.
2. [ESPAsyncDNSServer](https://github.com/devyte/ESPAsyncDNSServer) should be supported for ESP32<br>AutoConnect drives a DNS server for captive portal discovery. [ESPAsyncDNSServer](https://github.com/devyte/ESPAsyncDNSServer) does not support ESP32 officially, so some of it needs to be modified. (This is not too difficult)

Refs. https://github.com/Hieromon/PageBuilder/issues/12

If you just want to refresh the measured temperature display, I recommend using WebSocket or Ajax to asynchronize the web page. AutoConnect uses Ajax to asynchronously update the OTA progress bar. It will be helpful to you.

- Client side (Web page) JavaScript<br>Create an XMLHttpRequest object and register an event handler. Updates the progress bar placed on the page with the value (which includes the update size of the upload BIN file) notified in the progress event. <br>
https://github.com/Hieromon/AutoConnect/blob/e98018686175e836ed83cc98bbd09d4fbbd71ba9/src/AutoConnectUpdatePage.h#L41-L45

- Server side (viz. A part of the Sketch)<br>Returns the uploaded file size in response to a progress inquiry from the client.<br>
https://github.com/Hieromon/AutoConnect/blob/e98018686175e836ed83cc98bbd09d4fbbd71ba9/src/AutoConnectUpdate.cpp#L563-L564<br>https://github.com/Hieromon/AutoConnect/blob/e98018686175e836ed83cc98bbd09d4fbbd71ba9/src/AutoConnectUpdate.cpp#L584

- Server size (viz. A part of the Sketch)<br>Register the above response handler as a Web Server native handler.
https://github.com/Hieromon/AutoConnect/blob/e98018686175e836ed83cc98bbd09d4fbbd71ba9/src/AutoConnectUpdate.cpp#L183-L184
